### PR TITLE
add noise robustness experiment in multi_column_convergence

### PR DESCRIPTION
**Pull request by ywcui1990**
@subutai  Please review

* Add noise robustness experiment
* Add accuracy metrics: specificity, sensitivity and plain accuracy
* Compare column pooler with BOW classifier

Result 1 Number of touches vs. amount of noise
(I am not sure why having more columns give rise to worse noise robustness.)
image

Result 2 Accuracy vs. number of touches with different amount of noise
image

Result 3 Compare column pooler with BOW classifier with/without location
image

### CODE REVIEW

File: htmresearch/frameworks/layers/l2_l4_inference.py [hunk 1]
```diff
@@ -755,6 +760,24 @@ def getDefaultL4Params(self, inputSize, numInputBits):
       activationThreshold = int(numInputBits * .6)
       minThreshold = activationThreshold
 
+    # if self.config["L4RegionType"] == "py.ExtendedTMRegion":
```
Comment (subutai) [inline htmresearch/frameworks/layers/l2_l4_inference.py]: Why have commented lines? Either delete or keep.

File: htmresearch/frameworks/layers/l2_l4_inference.py [hunk 2]
```diff
@@ -767,10 +790,16 @@ def getDefaultL4Params(self, inputSize, numInputBits):
       "basalPredictedSegmentDecrement": 0.0,
       "apicalPredictedSegmentDecrement": 0.0,
       "activationThreshold": activationThreshold,
+      "reducedBasalThreshold": int(activationThreshold*0.6),
+    # int(activationThreshold*0.8),
```
Comment (subutai) [inline htmresearch/frameworks/layers/l2_l4_inference.py]: Why have commented lines? Either delete or keep.

File: htmresearch/frameworks/layers/l2_l4_inference.py [hunk 3]
```diff
@@ -767,10 +790,16 @@ def getDefaultL4Params(self, inputSize, numInputBits):
...
+    #   "reducedBasalThreshold": activationThreshold-1,# int(activationThreshold*0.8),
       "sampleSize": sampleSize,
-      "implementation": "ApicalTiebreakCPP",
+      "implementation": "ApicalTiebreak",
       "seed": self.seed
     }
+    # else:
```
Comment (subutai) [inline htmresearch/frameworks/layers/l2_l4_inference.py]: Should we still have this exception?

File: projects/l2_pooling/capacity_test.py [hunk 4]
```diff
@@ -309,7 +309,7 @@ def testOnSingleRandomSDR(objects, exp, numRepeats=100, repeatID=0):
     # Only set to 1 iff target object is the lone max overlap index.  Otherwise
     # the network failed to conclusively identify the target object.
     outcome[i] = 1 if maxOverlapIndices == [targetObject] else 0
-
+    print numActiveL2Cells
```
Comment (subutai) [inline projects/l2_pooling/capacity_test.py]: Do we need this print statement?

File: projects/l2_pooling/capacity_test.py [hunk 5]
```diff
@@ -477,7 +477,7 @@ def runCapacityTest(numObjects,
     numLocations=numLocations,
     numFeatures=numFeatures
   )
-
+  print "networkType: ", networkType
```
Comment (subutai) [inline projects/l2_pooling/capacity_test.py]: Do we need this print statement?

File: projects/l2_pooling/multi_column_convergence.py [hunk 6]
```diff
@@ -94,6 +94,42 @@ def averageConvergencePoint(inferenceStats, prefix, minOverlap, maxOverlap,
   return convergenceSum/len(inferenceStats)
 
 
+
+def averageConvergencePointNew(numActiveL2cells, minOverlap, maxOverlap):
```
Comment (subutai) [inline projects/l2_pooling/multi_column_convergence.py]: @lscheinkman We should refactor this to use the common function you created in https://github.com/numenta/htmresearch/pull/793

File: projects/l2_pooling/multi_column_convergence.py [hunk 7]
```diff
@@ -357,22 +519,37 @@ def runExperimentPool(numObjects,
...
+                           "locationNoise": locationNoise,
+                           "featureNoise": featureNoise,
+                           "enableFeedback": feedback,
+                           "numAmbiguousLocations": ambiguousLocations,
+                           "numInferenceRpts": numInferenceRpts,
+                           "l2Params": l2Params,
+                           "l4Params": l4Params
+                           }
                 )
+  if numWorkers > len(args):
```
Comment (subutai) [inline projects/l2_pooling/multi_column_convergence.py]: Don't think this is required.

File: projects/thing_classification/thing_convergence.py [hunk 8]
```diff
@@ -79,7 +79,7 @@ def getL2Params():
     "activationThresholdDistal": 13,
     "sampleSizeDistal": 30,
     "connectedPermanenceDistal": 0.5,
-    "distalSegmentInhibitionFactor": 1.001,
+    "distalSegmentInhibitionFactor": 0.9999,
```
Comment (subutai) [inline projects/thing_classification/thing_convergence.py]: Is this required?

**Comment by ywcui1990** [comment]
@subutai I have cleaned up unnecessary print statement and comments per your suggestions.

**Pull request merged**